### PR TITLE
feat #304: Extend the CompositeRank to allow for groups

### PR DESCRIPTION
**Pull request by mayurkmmt**
## Summary

- Extends `CompositeRank` to support group-based ranking via a new `group_col` parameter (string or list of strings)
- Ranks are now calculated independently within each group when `group_col` is specified
- Adds input validation for `group_col` (type checking, missing columns, empty list)
- Refactors `__init__` into smaller private methods (`_validate_group_col`, `_process_rank_columns`, `_parse_column_spec`, `_create_window`, `_create_composite_ranking`)

## Breaking change (bug fix)

`ignore_ties=True` now produces **1-based ranks** (1, 2, 3, ...) instead of **0-based ranks** (0, 1, 2, ...), consistent with `ignore_ties=False` behavior.

The original code had an incorrect comment claiming `ibis.row_number()` is 1-based — it is actually [0-based](https://ibis-project.org/reference/expression-generic#ibis.expr.types.relations.Table.row_number), same as `ibis.rank()`. This meant `ignore_ties=True` and `ignore_ties=False` produced ranks on different scales. Both modes now add `+ 1` to produce consistent 1-based ranks.

## Test plan

- [x] Existing tests for global ranking continue to pass
- [x] New tests for group-based ranking with all aggregation functions (mean, sum, min, max)
- [x] Tests for `ignore_ties=True` with group-based ranking verify unique 1-based ranks
- [x] Tests for invalid `group_col` inputs (wrong type, missing columns, empty list)
- [x] Tests for edge cases: single group, NULL values in group column, multi-column grouping
- [x] Tests for ibis table input with group-based ranking

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Comment by mayurkmmt** [comment]
@claude /review-pr-tests

File: tests/analysis/test_composite_rank.py [hunk 1]
```diff
@@ -204,3 +204,351 @@ def test_df_cache(self, simple_df):
...
+        # Verify margin ranks within groups
+        # Electronics margin ranks: product 2 (25%) = 1, product 1 (20%) = 2, product 3 (15%) = 3
+        assert electronics.loc[electronics["product_id"] == electronics_product_2, "margin_pct_rank"].iloc[0] == rank_1
+        assert electronics.loc[electronics["product_id"] == electronics_product_1, "margin_pct_rank"].iloc[0] == rank_2
+        assert electronics.loc[electronics["product_id"] == electronics_product_3, "margin_pct_rank"].iloc[0] == rank_3
+
+        # Apparel margin ranks: product 5 (35%) = 1, product 4(30%) = 2, product 6 (25%) = 3
+        assert apparel.loc[apparel["product_id"] == apparel_product_2, "margin_pct_rank"].iloc[0] == rank_1
+        assert apparel.loc[apparel["product_id"] == apparel_product_1, "margin_pct_rank"].iloc[0] == rank_2
+        assert apparel.loc[apparel["product_id"] == apparel_product_3, "margin_pct_rank"].iloc[0] == rank_3
```
Comment (murray-ds) [inline tests/analysis/test_composite_rank.py]: Rather than doing all these ilocs, etc can you please just create an expected results dataframe

File: tests/analysis/test_composite_rank.py [hunk 2]
```diff
@@ -204,3 +204,351 @@ def test_df_cache(self, simple_df):
...
+        electronics_product_1 = 1
+        electronics_product_2 = 2
+        electronics_product_3 = 3
+        apparel_product_1 = 4
+        apparel_product_2 = 5
+        apparel_product_3 = 6
+
+        rank_1 = 1
+        rank_2 = 2
+        rank_3 = 3
```
Comment (murray-ds) [inline tests/analysis/test_composite_rank.py]: These can be removed if you do an expected results dataframe

File: pyretailscience/analysis/composite_rank.py [hunk 3]
```diff
@@ -115,6 +115,7 @@ def __init__(
         rank_cols: list[tuple[str, str] | str],
         agg_func: str,
         ignore_ties: bool = False,
+        group_col: str | None = None,
```
Comment (murray-ds) [inline pyretailscience/analysis/composite_rank.py]: group_col should be able to be a list[str] or a list. Can you make that change please.

File: pyretailscience/analysis/composite_rank.py [hunk 4]
```diff
@@ -153,46 +162,115 @@ def __init__(
...
-                if len(col_spec) != 2:  # noqa: PLR2004 - Error message below explains the value
-                    msg = (
-                        f"Column specification must be a string or a tuple of (column_name, sort_order). Got {col_spec}"
-                    )
+        self._validate_group_col(group_col, df)
+        rank_mutates = self._process_rank_columns(rank_cols, df, group_col, ignore_ties)
+        df = df.mutate(**rank_mutates)
+        self.table = self._create_composite_ranking(df, rank_mutates, agg_func)
+
+    def _validate_group_col(self, group_col: str | list[str] | list | None, df: ibis.Table) -> None:
```
Comment (murray-ds) [inline pyretailscience/analysis/composite_rank.py]: List shouldn't be there. It should always be a list of strings
```suggestion
    def _validate_group_col(self, group_col: str | list[str] | None, df: ibis.Table) -> None:
```
